src/nxt_otel.c

```diff
@@ -0,0 +1,403 @@
+    nxt_log(task, NXT_LOG_ERR, "otel error condition");
+
+    /* assumable at time of writing that there is no
+     * r->otel->trace to leak. This state is only set
+     * in cases where trace fails to generate or is missing
+     */
+}
+
+
+void
+nxt_otel_trace_and_span_init(nxt_task_t *task, nxt_http_request_t *r)
+{
+    r->otel->trace =
+        nxt_otel_rs_get_or_create_trace(r->otel->trace_id);
+    if (r->otel->trace == NULL) {
+        nxt_log(task, NXT_LOG_ERR, "error generating otel span");
+        nxt_otel_state_transition(r->otel, NXT_OTEL_ERROR_STATE);
+        return;
+    }
+
+    nxt_otel_state_transition(r->otel, NXT_OTEL_HEADER_STATE);
+}
+
+
+void
+nxt_otel_test_and_call_state(nxt_task_t *task, nxt_http_request_t *r)
+{
+    if (r == NULL || r->otel == NULL) {
+        return;
+    }
+
+    switch (r->otel->status) {
+    case NXT_OTEL_UNINIT_STATE:
+      return;
+    case NXT_OTEL_INIT_STATE:
+        nxt_otel_trace_and_span_init(task, r);
+        break;
+    case NXT_OTEL_HEADER_STATE:
+        nxt_otel_span_add_headers(task, r);
+        break;
+    case NXT_OTEL_BODY_STATE:
+        nxt_otel_span_add_body(r);
+        break;
+    case NXT_OTEL_COLLECT_STATE:
+      nxt_otel_span_collect(task, r);
+        break;
+    case NXT_OTEL_ERROR_STATE:
+        nxt_otel_error(task, r);
+        break;
+    }
+}
+
+
+// called in nxt_http_request_error
+void
+nxt_otel_request_error_path(nxt_task_t *task, nxt_http_request_t *r) {
+  if (r->otel->trace == NULL) {
+        return;
+    }
+
+    // response headers have been cleared
+    nxt_otel_propagate_header(task, r);
+
+    // collect span immediately
+    if (r->otel) {
+        nxt_otel_state_transition(r->otel, NXT_OTEL_COLLECT_STATE);
+    }
+    nxt_otel_test_and_call_state(task, r);
+}
+
+
+nxt_int_t
+nxt_otel_parse_traceparent(void *ctx, nxt_http_field_t *field, uintptr_t data)
+{
+    char               *copy;
+    nxt_http_request_t *r;
+
+    /* For information on parsing the traceparent header:
+     * https://www.w3.org/TR/trace-context/#traceparent-header
+     * A summary of the traceparent header value format follows:
+     * Traceparent: "$a-$b-$c-$d"
+     *   a. version (2 hex digits) (ff is forbidden)
+     *   b. trace_id (32 hex digits) (all zeroes forbidden)
+     *   c. parent_id (16 hex digits) (all zeroes forbidden)
+     *   d. flags (2 hex digits)
+     */
+
+    r = ctx;
+    if (field->value_length != NXT_OTEL_TRACEPARENT_LEN) {
+        goto error_state;
+    }
+
+    /* strsep is destructive so we make a copy of the field
+     */
+    copy = nxt_mp_zalloc(r->mem_pool, field->value_length + 1);
+    if (copy == NULL) {
+        goto error_state;
+    }
+    memcpy(copy, field->value, field->value_length);
+
+    r->otel->version = (u_char *) strsep(&copy, "-");
+    r->otel->trace_id = (u_char *) strsep(&copy, "-");
+    r->otel->parent_id = (u_char *) strsep(&copy, "-");
+    r->otel->trace_flags = (u_char *) strsep(&copy, "-");
```

Comment:
This does give me pause for thought.

strsep(3) modifies it's first argument (copy), this is a problem with the following code

```c
char str[] = "Hello World";
char *s = strdup(str);

p = strsep(&s, " ");
...
free(s);
```

As we've just leaked the memory for s, but perhaps worse, introduced some undefined behaviour from free(2)'ing a now invalid pointer.

Yes, the above code may _seem_ to run OK...

In the above case what you need to do is introduce another pointer, and set that to s., you then use this new pointer in the strsep(3) call leaving s unchanged.

Our memory allocator may handle this, or I suspect this is a ticking time bomb.

The safest thing would be to just introduce a temporary pointer for use in strsep(3).

Seeing as this memory is only needed temporarily here, I'd just use nxt_strdup() rather than nxt_mp_zalloc() and then use nxt_free() (on the strdup(3)'d pointer) when finished.

Reply:
The memory is stored in the mempool. 
The r->otel fields are not free'd individually. 
This character buffer is not free'd separately from its mempool either.
This memory is not needed only temporarily, it is used during span creation far later in the HTTP request handling chain. 

This memory will be safely cleaned when the memory pool is deallocated.

Reply:
Yes, you are right about it being needed beyond the scope of this function.

As to my concern about `copy` no longer being 'valid', having checked, it's not an issue in this case.